### ISSUE TITLE
LightGBM: A Highly Efficient Gradient Boosting Decision Tree

### ISSUE THREAD

**Issue by ANGHOOO** [ANGHOOO/PaperReview#1]
Paper : https://proceedings.neurips.cc/paper_files/paper/2017/file/6449f44a102fde848669bdd9eb6b76fa-Paper.pdf
Github : https://github.com/microsoft/LightGBM/tree/master

### 논문개요
Gradient Boosting Decision Tree의 경우 인기있는 머신러닝 알고리즘이지만, 모든 Feature, 모든 data instances의 scan을 통해 Gradient를 측정해야 하기 때문에 계산 복잡성이 feature, instance의 수에 비례 $O(data * instances)$하게 되어 데이터의 크기가 커질수록 시간이 많이 소요된다는 점을 지적하였다. **feature의 차원**과 **data size**가 커지면 **efficiency**와 **scalability**의 관점에서 여전히 만족스럽지 못하기 때문에 이러한 한계점을 개선하기 위해 본 논문에서는 **Gradient-Based One Side Sampling**(GOSS)와 **Exclusive Feature Bundling**(EFB) 2가지 방식을 제안한다.
GOSS 방법은 gradient가 큰 data instances들을 통해 information gain을 추정하기 위한 방법으로 data instances들을 줄이는 방법이다.
EFB 방법은 상호 배타적인(mutually exclusive) feature들을 묶음으로써 feature들의 수를 줄이는 방법이다.
위의 2가지 방법을 통해 LightGBM은 기존 GBDT 모델에 비해 거의 비슷한 정확도를 달성하는데 최대 20배 이상 훈련 속도를 향상시켰다.

---

### 기존문제
기존 GBDT 모델을 학습하는데 gradient 계산을 위해 모든 data instaces와 feature들을 scan해야 한다는 점에서 계산 효율성을 지적한다.

---

### Main Contribution
- GOSS : Gradient-Based One Side Sampling(GOSS)는 gradient를 기준으로 instance들을 내림차순으로 정렬한 뒤, down sampling하는 방법이다. gradient가 작은 데이터 샘플들을 버려버리면 데이터 분포가 변화해서 학습된 모델의 정확도를 해칠 수 있기 때문에 GOSS 방법을 통해 해결하고자 하였다.

- EFB : EFB는 Feature Bundling을 통해 복잡도를 더욱 낮추는 방법으로 GOSS + EFB를 통해 기존 $O(data * feature)$에서 $O(down sampled data * feature bundle)$으로 복잡도를 낮추어 더욱 빠른 연산이 가능하게 하였다. EFB는 번들로 묶을 수 있는 **Feature를 식별**하는 단계와 **Feature들을 병합**하는 단계, 총 2단계로 구성된다.

- lgb_baseline 모델과 비교해서 LightGBM 모델은 Allstate, Flight Delay, LETOR, KDD10, KDD12 데이터셋에 대해서 각각 21x, 6x, 1.6x, 14x, 13x 더 빠른 학습 속도를 보여준다.

---

### Method

본 논문에서 GOSS 방법을 통해 down sampling 하는 방식은 다음의 수도코드와 같다.
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/94c87bbc-9541-44fa-b3f4-d5b121519ef1" width="400" height="300"/>

- **sorted** : $I$개의 instance를 gradient 기반으로 내림차순 정렬
- **topSet** : sorted된 리스트에서 topN개 만큼 추출 ($a$ : gradient가 큰 데이터에 대한 샘플링 비율, $len(I)$ : training data의 개수) 
- **randSet** : topN개를 제외한 데이터 샘플에서 randN개 만큼 추출 ($randN = b * len(I)$, $b$ : gradient가 작은 데이터에 대한 샘플링 비율)
- **usedSet** : topSet과 randSet을 합쳐서 다음의 학습에 사용

본 논문에서 EFB 방법을 통해 feature의 수를 줄이는 방식은 다음의 수도코드와 같다.
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/abeb2a35-afee-4e4e-a484-e86117ab447a" width="400" height="300"/>

EFB는 번들로 묶을 수 있는 Feature를 식별하는 단계와 Feature들을 병합하는 단계 총 2단계로 구성된다.
일반적으로 Feature가 많을 수록 차원이 높아지면서 0의 값들이 많아지게되고 Sparse한 데이터가 구성된다.
각 Feature들이 서로 동시에 0이 아닌 값(mutually exclusive)을 가질 확률이 낮아지기 때문에 정보손실을 최대한 줄이면서 Feature들을 합치는 방법이 EFB 방법의 핵심 아이디어다.
Feature의 값이 모두 0일 경우 Bundle의 손실없이 0으로 둘 수 있기에, 원본 데이터의 손실이 적은 접근이 가능하고 따라서, Bundling을 위해서는 각 Feature간에 0이 아닌 값을 가지는 경우에 집중한 효율적인 방법을 필요로 한다.

- **Part1 of EFB : 번들로 묶을 수 있는 Feature 식별**
1. Feature간에 0이 아닌 값을 함께 가지는 경우를 Conflict라고 하며, 각 Feature들 사이의 Conflict 수를 집계한다.
2. 집계된 각 Feature간의 Conflict의 총합을 계산하고 이렇게 Feature마다 구성된 값을 Degree로 명명.
3. 집계된 Degree들을 기준으로 Feature들을 내림차순으로 정렬한다 (Feature들을 묶고 값을 반환할 때, 값의 손실을 줄이기 위해 Conflict가 큰 Feature들은 적게 묶고 Conflict가 작은 Feature들은 많이 묶이도록 하기 위함이다.)
4. 각 Bundle에서 허용하는 최대 Conflict count를 정의한다. (최대 Conflict count가 클수록 더욱 많은 Feature가 묶이고 Bundle은 줄어들면서, 학습시간은 감소하지만 값의 손실은 증가)
5. 이후 논문의 수도코드에 나와있듯이 for loop을 순회하면서 진행하는 방식으로 번들로 묶을 수 있는 Feature를 식별

- **Part2 of EFB : Feature 병합**
1. Bundle내에서 큰 Degree를 가진 Feature가 기준 Feature로 선정 (다른 Feature와 가장 많은 Conflict를 발생한 Feature, Bundle내의 값들을 조합해 하나의 값으로 변환하는데에 있어 가장 큰 손실을 초래할 수 있음, 따라서 해당 feature를 기준으로 최소한의 손실로 값을 변환할 수 있는 기법을 사용)
2. 기준 Feature의 가장 많은 Conflict가 발생한 경우(최댓값)을 찾고, Bundle내 값의 변환에 사용할 Offset으로 지정 (기준 Feature가 0의 값을 가지고 Bundle내의 다른 Feature들이 0이 아닐 경우, 다른 Feature의 값에 기준 Feature가 가질 수 있는 최대 Offset을 추가하면서 사용, 해당 열의 Data에서는 Conflict가 0이지만, 다른 데이터에 대해서 Offset만큼 기준 Feature에서 Conflict가 발생할 수 있기 때문)
3. Conflict가 일어난 경우, 기준 Feature의 값을 Bundle값으로 지정 (값의 손실을 최소화 하기 위함)
4. Conflict가 일어나지 않은 경우를 계산 (기준이 아닌 Feature의 값이 0이 아닌 경우, 기준이 아닌 Feature의 값에 Offset을 더하여 Bundle의 값으로 지정, 기준이 아닌 Feature의 값이 0인 경우, 충돌은 기준 Feature에서만 발생했으므로 기준 Feature의 값을 Bundle값으로 지정)

---

### Experiment
본 논문에서는 다음과 같은 데이터셋들을 사용하였다.
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/c9a3907f-2a90-45b5-b1d0-e0948b64e373" width="600" height="200"/>

다른 모델들과 비교했을 때 lightGBM 모델의 학습속도가 월등히 빠른 것을 확인할 수 있다.
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/223a504b-aaae-4da4-967b-63440b193cec" width="600" height="200"/>

다른 모델들과 비교했을 때 성능은 거의 동일한 것을 볼 수 있지만 동일한 성능을 내는데까지의 학습속도는 위에서 볼 수 있듯이 차이가 난다.
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/ed3af7f3-9367-4090-970b-954fa8f1468a" width="600" height="200"/>

LightGBM 모델이 베이스라인과 같은 성능을 내는데까지 훨씬 빠른 속도로 학습이 된다는 것을 볼 수 있다.
lgb_baseline 모델과 비교해서 LightGBM 모델은 Allstate, Flight Delay, LETOR, KDD10, KDD12 데이터셋에 대하여 각각 21x, 6x, 1.6x, 14x, 13x 더 빠른 것을 확인할 수 있다.
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/c6a44d15-fa15-4cfb-a1d9-a749f41ae33b" width="600" height="400"/>

모든 데이터셋에 대해 LightGBM 모델은 거의 비슷한 성능을 보인다. 이는 GOSS와 EFB 방법이 정확도는 해치지 않으면서 속도향상 측면에서는 크게 기여하는 것을 의미한다.
<img src="https://github.com/ANGHOOO/PaperReview/assets/103275370/bb242238-7b6f-491c-8106-ab22853c2282" width="600" height="200"/>

---

### Personal Thought
학습 효율을 높이기 위해 제시한 GOSS, EFB 방법들이 매우 흥미로웠다.
논문의 길이도 짧고 문장들도 읽기 쉽게 구성되어 있어서 논문을 읽는 데에는 크게 어려움이 없었다. 또한 방법론 자체도 크게 어려운 내용은 아니었다고 생각된다. (미리 공부를 한 상태에서 논문을 읽었기 때문인 것 같지만...)
하지만 논문의 수도코드 부분에서 생략된 부분도 있었고 오탈자도 있었기 때문에 단순 수도코드만 보고 이해하기는 쉽지 않았다.

논문 리뷰 full version
https://anghoo-ai.tistory.com/32
